File: 백준/Silver/28215. 대피소/대피소.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var in = bufio.NewScanner(os.Stdin)

func nextInt() int {
	in.Scan()
	r := 0
	for _, c := range in.Bytes() {
		r *= 10
		r += int(c - '0')
	}
	return r
}

var N, K int
var list []int
var ans = math.MaxInt32
var location [][]int

func f(k, i int) {
	if k == K {
		res := 0
		min := make([]int, N)
		for j := range min {
			min[j] = math.MaxInt32
		}

		for _, shelter := range list {
			a := location[shelter][0]
			b := location[shelter][1]
			for j := 0; j < N; j++ {
				x := abs(a - location[j][0])
				y := abs(b - location[j][1])
				min[j] = minInt(min[j], x+y)
			}
		}
		for _, n := range min {
			res = maxInt(res, n)
		}
		ans = minInt(ans, res)
		return
	}
	for j := i + 1; j < N; j++ {
		list[k] = j
		f(k+1, j)
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	in.Split(bufio.ScanWords)
	N = nextInt()
	K = nextInt()

	list = make([]int, K)
	location = make([][]int, N)

	for i := 0; i < N; i++ {
		location[i] = make([]int, 2)
		location[i][0] = nextInt()
		location[i][1] = nextInt()
	}
	f(0, -1)
	fmt.Println(ans)
}

File: 백준/Silver/28215. 대피소/대피소.java
import java.io.*;
import java.util.*;
public class Main{
    static int N,K,list[], ans =Integer.MAX_VALUE, location[][],min[];

    static void f(int k,int i) {
        if (k == K) {
            int res = 0;
            min = new int[N];
            Arrays.fill(min, Integer.MAX_VALUE);
            for (int shelter:list){
                int a = location[shelter][0], b = location[shelter][1];
                for (int j=0;j<N;j++){
                    int x = Math.abs(a-location[j][0]), y = Math.abs(b-location[j][1]);
                    min[j] = Math.min(min[j], x+y);
                }
            }
            for (int n:min) res = Math.max(res,n);
            ans = Math.min(ans,res);
            return;
        }
        for (int j = i+1; j < N; j++) {
            list[k] = j;
            f(k+1,j);
        }
    }

    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        StringBuilder ans = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        list = new int[K];
        location = new int[N][2] ;

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            location[i][0] = Integer.parseInt(st.nextToken());
            location[i][1] = Integer.parseInt(st.nextToken());
        }
        f(0,-1);
        System.out.println(ans);
    }
}